### PR TITLE
[Documentation] Python Automation script and workflow for github wiki

### LINKED ISSUE #133: [Documentation] Automate a Github wiki from repo READMEs
# Objective
Create a Github Wiki for Pocket V1

# Origin Document
The main page for https://github.com/pokt-network/pocket contains links to several READMEs spread through the network.

There are a few problems with the current documentation:
1. Some READMEs are still hidden throughout the repo
2. Getting a high-level view of all the documentation is difficult
3. There is no summary of the README
4. Finding the appropriate / correct README is hard

## Starting point

A good starting point would be to view the structure of https://github.com/pokt-network/pocket.

In order to find all the existing READMEs, one could execute `find . -name "README.md" | grep -v "vendor"` which outputs the following:
 

```
./shared/types/genesis/README.md
./shared/README.md
./docs/roadmap/README.md
./docs/development/README.md
./docs/deps/README.md
./docs/build/README.md
./README.md
./persistence/README.md
./persistence/pre_persistence/README.md
./p2p/pre2p/README.md
./p2p/README.md
./utility/doc/README.md
```

## References

A good example or reference of this is the [RockdDB Wiki Page](https://github.com/facebook/rocksdb/wiki), which currently looks like this:

Screen Shot 2022-07-28 at 11 23 47 AM

# Goals / Deliverables
- [ ] Github Wiki is added to the core repo
- [ ] A process is defined on how to keep it up to date (this can be automated or defined via a manual set of step)

# Optional Goals
- [ ] Provide a summary of each README

# Non-goals
- Adding new documentation to the rest of the repo

---

Creator: @Olshansk 
Co-Owners: ???

### PR DESCRIPTION
**Pull request by profishional**
### Motivation

Create a [GitHub wiki](https://docs.github.com/en/communities/documenting-your-project-with-wikis/about-wikis) (a separate git repo relative to this repo) for the documentation across this GitHub repo. GitHub wikis have better visibility and navigation for users.

### Background

GitHub wikis lack version control and are entirely separate relative to this source repository. To solve both of these issues, we use the documentation/markdown files from the main repo to update the Github wiki repository.

To maximize the benefit of the wiki, a hierarchical view is created in a sidebar. However, the mapping of the markdowns files to the sidebar poses some challenges. 

The PR is an attempt at a very simple solution. A GitHub action that tracks changes to a set of files and triggers a python script. The script picks up pre-written comments in the set of files to generate a sidebar. From the sidebars linking reference, it copies over the files from the main repo to the wiki repo.

### Pre-Requisites to Merge this in

- [ ] Create a repository secret. Based on the legacy access rules, tick the `repo` permission box.

### Outstanding issues

- [ ] Improve the format of the sidebar, so categories don't repeat. Rolling the main README of a category into the category heading so that the heading is also a link and not just text.
- [ ] Change the internal links of the Markdown files in the wiki to point to other paths in the wiki by either:
   1.  A regex change to the current wiki file names (aka wiki titles)
   2. Changing the wiki file mapping to align with the internal links.

## Issue

Fixes #133 

Remaining/new issues

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [x] Documentation
- [ ] Other 

## List of changes

- (Apologies in advance) Some formatting changes from auto linting of markdowns files. I don't like this as it has blown out the diff. I can try to remove if it's too messy.
- Added a GitHub action that tracks change to `.md` files, sets env variables and runs a python script
- Added a python script in a new directory called tools. Finds the path of `.md` files, maps them to the specified format described in the files, generates a GitHub wiki sidebar file and respective wiki files. Pulls the current wiki, syncs the files and pushes them. 
- Added comments at the bottom of `.md` files found from `find . -name "*.md" | grep -v -e "vendor" -e "app"` in the form of `<!-- GITHUB_WIKI: category/subcategory/.../file_name -->`.
- Brief description of maintaining the wiki format comment in the `docs/development/readme`
- Brief mention and linking to the wiki in the main readme.
 
## Testing

- [ ] `make develop_test`
- [ ] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`
d

Screenshot of output from example wiki on a personal fork: [example wiki](https://github.com/profishional/pocket/wiki)
image

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [x] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

**Comment by profishional** [comment]
Just FYI @Olshansk; I am not able to set the Assignee, label, priority, etc. while creating the pull request during or after submitting the draft PR.
This is an example of a cross repo PR that has the buttons:
image

This is what I see while trying to creating a cross repo PR for pocket:
image

This is what it looked like after creating the draft PR, usually the words become clickable links to set the assignee etc.:
image

Maybe there is a different way, but those are the ones I was familiar with.

File: tools/wiki_sync.py
Comment (Olshansk) [inline tools/wiki_sync.py:38]: Please add a comment on what kind of categorization this does

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: tools/wiki_sync.py
Comment (profishional) [reply tools/wiki_sync.py:35]: True, I was relying on the `find | grep` too much.

File: consensus/doc/CHANGELOG.md
Comment (profishional) [reply consensus/doc/CHANGELOG.md]: True, thought it would break the `changelog` search in the wiki - it doesn't

File: tools/wiki_sync.py
Comment (profishional) [reply tools/wiki_sync.py:32]: Sorry I don't follow. I presume it's about improving the `md` comment parsing but what is "s/wiki_..."?

File: docs/development/README.md
Comment (profishional) [reply docs/development/README.md:81]: So this is one of the things that my "auto format on save" does in vscode for`md` files. Hmm I'll go fix these up and just turn change my vscode stuff.

File: .github/workflows/wiki_sync_process.yml
Comment (Olshansk) [inline .github/workflows/wiki_sync_process.yml:11]: Leaving an AI on @Olshansk to do this. 

Cc @okdas  who's been helping with other secrets management in our repo

File: docs/development/README.md
Comment (Olshansk) [reply docs/development/README.md:81]: We should use `make` instead of `$ make`. I never had the chance to properly implement #257 yet so everyone's VSCode settings in this repo are identical.

File: persistence/docs/PROTOCOL_STATE_HASH.md
Comment (Olshansk) [inline persistence/docs/PROTOCOL_STATE_HASH.md:140]:
```suggestion
<!-- GITHUB_WIKI: persistence/state_hash_protocol -->
```

File: persistence/docs/README.md
Comment (Olshansk) [inline persistence/docs/README.md:23]:
```suggestion
<!-- TODO(#77): Implement proper SQL migrations -->
```

File: persistence/docs/README.md
Comment (Olshansk) [inline persistence/docs/README.md:179]:
```suggestion
- [ ] CLEANUP(#76): Review all the `gov_*.go` related files and simplify the code
```

Github automatically links to `#XXX`. Applies bellow

File: shared/docs/PROTOCOL_STATE_HASH.md
Comment (Olshansk) [inline shared/docs/PROTOCOL_STATE_HASH.md:177]:
```suggestion
<!-- GITHUB_WIKI: shared/state_hash_protocol -->
```

File: docs/development/README.md
Comment (Olshansk) [inline docs/development/README.md:186]:
```suggestion
The Action is triggered when there is a change to any Markdown file on the main branch (i.e. after merge) of the Repository. When triggered, environment variables are set for a Python script that updates the Github Wiki Repository based on Pocket Repository source code.
```

File: docs/development/README.md
Comment (Olshansk) [inline docs/development/README.md:190]:
```suggestion
The script finds the relevant Markdown files in the repository and organizes them for the Wiki Repository. Currently, the find command is filtered to exclude the `./app` and `./vendor` directories. Based on the list of `.md` file paths, it maps the formatting spec from [above](##Maintaining-Documentation) to some information about the file. Using the map, it creates a Sidebar file which Github uses as a Table of Contents for the wiki, and also copies over the files with titles linking to the Sidebar from the Pocket repo.
```

File: docs/development/README.md
Comment (Olshansk) [inline docs/development/README.md:191]: Bonus points for adding a screenshot of the sidebar

**Review by Olshansk — CHANGES_REQUESTED** [review]
Last round of comments and one action-item on me to configure the secrets appropriately.

File: docs/development/README.md
Comment (profishional) [reply docs/development/README.md]: I just went and reverse all the formatting stuff because its annoying me, also makes it easier for me to review my own diff! I commit the ones you commented but may as well do them all (trailing spaces, lines, md format, tabs, issues/pr links, etc.) together at some point.

File: tools/wiki_sync.py
Comment (profishional) [reply tools/wiki_sync.py:50]: Ah yes did this in other parts. I like reduce indentations esp with bracketed languages

**Comment by Olshansk** [comment]
@okdas Could you help configure the secrets in our github repo to enable this via automation? See `.github/workflows/wiki_sync_process.yml`

**Comment by okdas** [comment]
> @okdas Could you help configure the secrets in our github repo to enable this via automation? See `.github/workflows/wiki_sync_process.yml`

I somehow lost access to this functionality, I'll work on either getting access or setting the secrets with the infra team.

**Comment by Olshansk** [comment]
> > @okdas Could you help configure the secrets in our github repo to enable this via automation? See `.github/workflows/wiki_sync_process.yml`
> 
> I somehow lost access to this functionality, I'll work on either getting access or setting the secrets with the infra team.

Updated the settings. Let me know if you've taken another look

File: tools/wiki_sync.py
Comment (okdas) [inline tools/wiki_sync.py:120]: It doesn't appear we can create a personal access token under organization, unless it is a [fine-grained PAT](https://github.blog/2022-10-18-introducing-fine-grained-personal-access-tokens-for-github/). I don't want us to utilize tokens that belong to particular people, and would rather use ones that are tied to `pokt-network` organization. The problem with fine-grained PAR, however, is it is not obvious [which permissions](https://docs.github.com/en/rest/overview/permissions-required-for-fine-grained-personal-access-tokens?apiVersion=2022-11-28) to add to allow to push to wiki. I am happy to research and figure out which permissions are actually needed, BUT I suspect there is a better way. Let's try this first, and if it doesn't work we can defer to fine-grained PAT.

From inside the GitHub action CI we should have all necessary permissions to interact with wiki that belongs to the same repo with `GITHUB_ACTOR` and `GITHUB_TOKEN` variables. While `GITHUB_ACTOR` environment variable is available in runtime, `GITHUB_TOKEN` is a secret that should be passed as an environment variable. I'll add a similar suggestion to `.github/workflows/wiki_sync_process.yml` to add it.

Feel free to move `os.environ` to separate variables, I just want to demonstrate how this pull CLI should look like:

```suggestion
        ["git", "pull", f"https://{os.environ["GITHUB_ACTOR"]}:{os.environ["GITHUB_TOKEN"]}@github.com/{owner}/{repo_name}.wiki.git"],
```

File: .github/workflows/wiki_sync_process.yml
Comment (okdas) [reply .github/workflows/wiki_sync_process.yml:11]: As per suggestion in python script, we need to add `GITHUB_TOKEN`, which is a token managed by GitHub Actions and can be used with `GITHUB_ACTOR` environment variable to access wiki via github.
```suggestion
  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} # TODO_IN_THIS_COMMIT: provision repo secret through Pocket account
```

File: tools/wiki_sync.py
Comment (okdas) [inline tools/wiki_sync.py:155]: Similarly, this change is necessary:

```suggestion
            f"https://{os.environ["GITHUB_ACTOR"]}:{os.environ["GITHUB_TOKEN"]}@github.com/{owner}/{repo_name}.wiki.git",
```

**Review by okdas — COMMENTED** [review]
Thank you for taking care of this @profishional!

I've explained why tokens might not be the best way to access wiki with an alternative solution in comments in code. From my research, this should work just fine! However, if something won't work as suspected, I'll research what permissions we are going to need for fine-grained PAT.

File: tools/wiki_sync.py
Comment (Olshansk) [reply tools/wiki_sync.py:120]: @okdas Appreciate the suggestions and explanation!

Wanted to ask: so should this just "work out of the box" if we make these changes or do we need to set up additional secrets in the repo settings?

Comment (okdas) [reply tools/wiki_sync.py:120]: > Wanted to ask: so should this just "work out of the box" if we make these changes or do we need to set up additional secrets in the repo settings?

Yes, it should! This is my understanding from a brief research – that `GITHUB_TOKEN`, which is a unique secret generated for each individual job run, is supposed to have access to pull/push wiki data for the same repo. This would be ideal, of course, so we don't have to manage additional secrets associated with people.

**Comment by Olshansk** [comment]
@profishional Let us know if you have time/capacity to tend to the changes we mentioned so we can get this PR over the finish line. If not, @okdas and I can push it ourselves

File: tools/wiki_sync.py
Comment (profishional) [reply tools/wiki_sync.py]: Ah interesting. This seems to have worked when I did it on my own fork. 
For reference if it needs to be reviewed: I used "Generate New Token (Classic)" and just gave it the whole "Repo" access - maybe it can be even more specific than that. 

Lmk how it goes.

**Review by Olshansk — APPROVED** [review]
@profishional LGTM!

Can you merge with main and resolve the merge conflicts? Since this branch is from your own repository, I don't have the ability to update it.

Screenshot 2023-02-02 at 11 52 31 AM

**Comment by profishional** [comment]
Ok I've solved the conflict. But can't merge "only those with write access to this repo can merge PR".

**Comment by Olshansk** [comment]
@profishional I can merge it for you. Can you update the changelogs first please? Again, I would do it on your behalf, but this branch isn't in our main repo.

Screenshot 2023-02-02 at 5 40 53 PM

**Comment by profishional** [comment]
I don't follow. You've merged it but you'd like me to update the Changelogs as another PR?

**Comment by Olshansk** [comment]
@profishional Sorry about that. I wrote the original message but then added an EDIT at the bottom:

> EDIT: Ignore. I will make an exception because of how minimal the changes in each repo are.

We'd be updating every changelog saying "Added a GITHUB WIKI tag", so the changelog diff would be greater than the actual diff, which I saw as unnecessary overhead. I made an exception and merged the changes as is.

I also just created a followup ticket that's adding it in the missing READMEs here: #488.

I think this is really awesome btw!! https://github.com/pokt-network/pocket/wiki

**Pull request merged**
